demos/conditionvariables/README.md

```diff
@@ -0,0 +1,50 @@
+# Condition Variables
+
+In this demo we explain the mechanism in condition variables.
+
+Condition variable is a synchronization tool which can block threads until a certain condition happens. 
+
+To show the concept behind it, we need to introduce the following:
+
+* `init_condition_variable` (in `src/user/condition_variable.c`)
+* `wait` (in `src/user/condition_variable.c`)
+* `signal` (in `src/user/condition_variable.c`)
+* `broadcast` (in `src/user/condition_variable.c`)
+
+## `init_condition_variable`
+
+This function initializes the condition variable pointed by `cv`.
+
+Before using functions like `wait()`, `signal()` and `broadcast()`, we must set up the variables with this function.
+
+## `wait`
+
+When we use this function, the thread will be blocked by the condition variable pointed by `cv`.
+
+After calling the function, the thread's `ThreadState` will be changed to `wait`. Since the scheduler only chooses the thread whose state is either `init`, `running` or `suspended`, the blocking thread will keep waiting until other threads signal through the same condition variable.
+
+## `signal`
+
+This function unblocks a thread that is blocked on the specified condition variable `cv`.
+
+`signal()` unlocks the thread by setting the state to `suspended`. So the thread can be chosen by the scheduler.
+
+If there exists a thread that can be waked up, the function returns `true`. When there has no thread can be signalled, it returns `false` as a consequence.
+
+## `broadcast`
+
+This function unblocks all threads blocked on the specified condition variable `cv`.
+
+`broadcast()` has a similar functionality to `signal()`. However, when we use `broadcast()`, we will wake up all the threads blocked by the condition variable while `signal()` only wakes up one of them.
+
+## Walkthrough
+
+From starter, the `setup()` function creates 5 threads that are set to run `waiter()` and initializes a condition variable called `cond_var`. After the creation, the `setup()` thread jumps to `signaller()` function.
+
+In `signaller()`, it first yields to check if the condition variable successfully blocks all other threads. Then the `signaller` thread uses `signal()` to wake 2 threads. The scheduler will choose the thread waked up by `cond_var` after the `signaller` thread yields.
+
+Once thread 1 and thread 2 finish their own routine, the `signaller` thread calls `broadcast()` and wakes up all threads except itself. After that, the `signaller` tries to call `signal()`, but since there is no thread blocked by `cond_var`, `signal()` returns `false`.
+
+Then `signaller` creates another thread called `final_signal` and waits for `cond_var`, which shows that a thread can wait for a condition variable which has once been signalled.
+
+When the `final_signal` thread calls `signal()` on `cond_var`, it finishes its execution. Since the `signaller` has no job to do, it exits after it is chosen to run and ends the demo.
```

Comment:
"Then the `final_signal` thread calls `signal()` on `cond_var` and finishes. Next `signaller` runs, it has no work left to do so exits and ends the demo."